### PR TITLE
Inline Edit Feature

### PR DESCRIPTION
**Pull request by hediet**

### CODE REVIEW

File: src/vscode-dts/vscode.proposed.inlineEdit.d.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+		/**
+		 * Provide inline edit for the given document.
+		 *
+		 * @param document The document for which the inline edit are computed.
+		 * @param context Additional context information about the request.
+		 * @param token A cancellation token.
+		 * @return An inline edit or a thenable that resolves to such. The lack of a result can be
+		 * signaled by returning `undefined` or `null`.
+		 */
+		provideInlineEdit(document: TextDocument, context: InlineEditContext, token: CancellationToken): ProviderResult<InlineEdit>;
```
Comment (hediet) [inline src/vscode-dts/vscode.proposed.inlineEdit.d.ts:74]: But this just returns one inline edit, right?

File: src/vscode-dts/vscode.proposed.inlineEdit.d.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+		/**
+		 * The new text for this edit.
+		 */
+		text: string;
+
+		/**
+		 * An optional range that will be replaced by the text of the inline edit.
+		 */
+		replaceRange?: Range;
```
Comment (hediet) [inline src/vscode-dts/vscode.proposed.inlineEdit.d.ts:21]: What if `replaceRange` does not agree with `position`? Maybe there should just be a required `replaceRange` (which can be empty)?

File: src/vs/code/electron-sandbox/workbench/workbench-dev.html [hunk 3]
```diff
@@ -56,6 +56,7 @@
 					domLineBreaksComputer
 					dompurify
 					editorGhostText
+					editorInlineEdit
```
Comment (hediet) [inline src/vs/code/electron-sandbox/workbench/workbench-dev.html:59]: Try to not touch this list. github.dev etc. has to be updated for this as well. Cannot you reuse `editorGhostText`?

File: src/vs/workbench/api/common/extHostApiCommands.ts [hunk 4]
```diff
@@ -493,6 +493,18 @@ const newCommands: ApiCommand[] = [
 		],
 		ApiCommandResult.Void
 	),
+	// --- multi ghost text
+	new ApiCommand(
+		'vscode.showMultiGhostText', '_showMultiGhostText', 'Show ghost text for multiple ranges',
```
Comment (hediet) [inline src/vs/workbench/api/common/extHostApiCommands.ts:498]: This should go

File: src/vs/workbench/services/actions/common/menusExtensionPoint.ts [hunk 5]
```diff
@@ -364,6 +364,13 @@ const apiMenus: IAPIMenu[] = [
 		supportsSubmenus: false,
 		proposed: 'inlineCompletionsAdditions'
 	},
+	{
+		key: 'editor/inlineEdit/actions',
+		id: MenuId.InlineEditActions,
+		description: localize('inlineEdit.actions', "The actions shown when hovering on an inline edit"),
+		supportsSubmenus: false,
+		// proposed: 'inlineEdit'
```
Comment (hediet) [inline src/vs/workbench/services/actions/common/menusExtensionPoint.ts:372]: wip

File: src/vs/editor/common/config/editorOptions.ts [hunk 6]
```diff
@@ -428,6 +428,7 @@ export interface IEditorOptions {
 	 */
 	suggest?: ISuggestOptions;
 	inlineSuggest?: IInlineSuggestOptions;
+	inlineEdit?: IInlineEditOptions;
```
Comment (hediet) [inline src/vs/editor/common/config/editorOptions.ts:431]: Maybe this should be called "experimentalInlineEdit", to make clear that this all might be removed/changed.

File: src/vscode-dts/vscode.proposed.inlineEdit.d.ts (hunk 1 above)
Comment (Krzysztof-Cieslak) [reply src/vscode-dts/vscode.proposed.inlineEdit.d.ts:74]: Yes, we decided to initially go with the single edit at the time.

File: src/vs/workbench/api/common/extHostApiCommands.ts (hunk 4 above)
Comment (Krzysztof-Cieslak) [reply src/vs/workbench/api/common/extHostApiCommands.ts]: Done

File: src/vs/editor/common/config/editorOptions.ts (hunk 6 above)
Comment (Krzysztof-Cieslak) [reply src/vs/editor/common/config/editorOptions.ts]: ✅

File: src/vs/editor/standalone/browser/standaloneLanguages.ts [hunk 7]
```diff
@@ -671,6 +671,11 @@ export function registerInlineCompletionsProvider(languageSelector: LanguageSele
 	return languageFeaturesService.inlineCompletionsProvider.register(languageSelector, provider);
 }
 
+export function registerInlineEditProvider(languageSelector: LanguageSelector, provider: languages.InlineEditProvider): IDisposable {
```
[the remark this answers is not in the document]
Comment (Krzysztof-Cieslak) [reply src/vs/editor/standalone/browser/standaloneLanguages.ts:674]: It seems to me that `monaco.d.ts` file is getting automatically updated with my proposed changes to the API, and then the type signature on the `createMonacoLanguagesAPI` function forces me to define this function

File: src/vs/code/electron-sandbox/workbench/workbench-dev.html (hunk 3 above)
Comment (Krzysztof-Cieslak) [reply src/vs/code/electron-sandbox/workbench/workbench-dev.html]: ✅

**Comment by hediet** [comment]
Generally, can you try to reuse as much as possible (withoung changing the original much) from the inline completions implementation?
At least the utils you could export and reuse, you don't have to copy them.

File: src/vscode-dts/vscode.proposed.inlineEdit.d.ts (hunk 2 above)
Comment (Krzysztof-Cieslak) [reply src/vscode-dts/vscode.proposed.inlineEdit.d.ts]: That's a good idea 👍 

I've changed the API just to be required `range` and removed `position`

**Comment by Krzysztof-Cieslak** [comment]
OK, I've tried to clean it up a bit. `utils.ts` and `ghostText.ts` are no longer duplicated, and I'm now reusing parts of `ghostTextWidget.ts` (the big remaining "pain" point here is the `GhostTextWidget` class itself, which is very similar in both implementations - I probably could look into refactoring it a bit to enable better code reuse, but it would require bigger changes in the `inlineCompletions` version)

File: src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts [hunk 8]
```diff
@@ -202,7 +203,8 @@ class AdditionalLinesWidget extends Disposable {
 	constructor(
 		private readonly editor: ICodeEditor,
 		private readonly languageIdCodec: ILanguageIdCodec,
-		private readonly lines: IObservable<{ targetTextModel: ITextModel; lineNumber: number; additionalLines: LineData[]; minReservedLineCount: number } | undefined>
+		private readonly lines: IObservable<{ targetTextModel: ITextModel; lineNumber: number; additionalLines: LineData[]; minReservedLineCount: number } | undefined>,
+		private readonly cls: string
```
Comment (hediet) [inline src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts:207]: Can you use a more descriptive value for this string and/or add docs?
Maybe just `className`, with docs being "the css class of the view-lines"?
Also, the value should be validated to prevent escaping issues (see DecorationOptions for validation examples).

File: src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+.monaco-editor .inline-edit {
+	font-family: ${fontFamily};
+}`);
+		}));
+
+		const backgroundRemoved = this._register(createStyleSheet2());
+		this._register(autorun(reader => {
+			const backgroundColoring = this._backgroundColoring.read(reader);
+			backgroundRemoved.dispose();
+			backgroundRemoved.setStyle(backgroundColoring ? `
```
Comment (hediet) [inline src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts:117]: Optional: Can you solve this via a class (`.inline-edit-remove.backgroundColoring`)? Generally, we try to avoid dynamic JS CSS manipulation.

File: src/vs/workbench/api/common/extHost.api.impl.ts [hunk 10]
```diff
@@ -624,6 +624,11 @@ export function createApiFactoryAndRegisterActors(accessor: ServicesAccessor): I
 				}
 				return extHostLanguageFeatures.registerInlineCompletionsProvider(extension, checkSelector(selector), provider, metadata);
 			},
+			registerInlineEditProvider(selector: vscode.DocumentSelector, provider: vscode.InlineEditProvider): vscode.Disposable {
+				//TODO: reneable proposed API check
+				// checkProposedApiEnabled(extension, 'inlineEdit');
```
[the remark this answers is not in the document]
Comment (hediet) [reply src/vs/workbench/api/common/extHost.api.impl.ts:629]: Still a blocker

File: src/vs/workbench/api/test/browser/extHostTextEditor.test.ts [hunk 11]
```diff
@@ -89,7 +89,7 @@ suite('ExtHostTextEditorOptions', () => {
 			$trySetSelections: undefined!,
 			$tryApplyEdits: undefined!,
 			$tryInsertSnippet: undefined!,
-			$getDiffInformation: undefined!
+			$getDiffInformation: undefined!,
```
Comment (hediet) [inline src/vs/workbench/api/test/browser/extHostTextEditor.test.ts:92]: This is an unnecessary change

File: src/vs/workbench/services/actions/common/menusExtensionPoint.ts (hunk 5 above)
Comment (hediet) [reply src/vs/workbench/services/actions/common/menusExtensionPoint.ts:390]: ping

File: src/vs/workbench/api/common/extHostTypes.ts [hunk 12]
```diff
@@ -4277,3 +4277,19 @@ export enum KeywordRecognitionStatus {
 }
 
 //#endregion
+
+//#region InlineEdit
+
+export class InlineEdit implements vscode.InlineEdit {
+	constructor(
+		public readonly text: string,
+		public readonly range: Range,
```
Comment (hediet) [inline src/vs/workbench/api/common/extHostTypes.ts:4286]: This does not properly implement the interface, as the interface is not readonly.

**Review by hediet — COMMENTED** [review]
See comments

File: src/vs/editor/common/config/editorOptions.ts [hunk 13]
```diff
@@ -4072,6 +4073,83 @@ class InlineEditorSuggest extends BaseEditorOption<EditorOption.inlineSuggest, I
...
+
+/**
+ * @internal
+ */
+export type InternalInlineEditOptions = Readonly<Required<IInlineEditOptions>>;
+
+class InlineEditorEdit extends BaseEditorOption<EditorOption.inlineEdit, IInlineEditOptions, InternalInlineEditOptions> {
+	constructor() {
+		const defaults: InternalInlineEditOptions = {
+			enabled: true,
```
[the remark this answers is not in the document]
Comment (Krzysztof-Cieslak) [reply src/vs/editor/common/config/editorOptions.ts]: ✅ 
We will override that with `configurationDefaults` on the extension side.

File: src/vs/workbench/api/test/browser/extHostTextEditor.test.ts (hunk 11 above)
Comment (Krzysztof-Cieslak) [reply src/vs/workbench/api/test/browser/extHostTextEditor.test.ts]: ✅

File: src/vs/workbench/api/common/extHost.api.impl.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (Krzysztof-Cieslak) [reply src/vs/workbench/api/common/extHost.api.impl.ts]: ✅

File: src/vs/workbench/services/actions/common/menusExtensionPoint.ts (hunk 5 above)
Comment (Krzysztof-Cieslak) [reply src/vs/workbench/services/actions/common/menusExtensionPoint.ts]: ✅

File: src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+
+	public static readonly cursorAtInlineEditKey = 'cursorAtInlineEdit';
+	public static readonly cursorAtInlineEditContext = new RawContextKey<boolean>(InlineEditController.cursorAtInlineEditKey, false);
+	private _isCursorAtInlineEditContext = InlineEditController.cursorAtInlineEditContext.bindTo(this.contextKeyService);
+
+	public static get(editor: ICodeEditor): InlineEditController | null {
+		return editor.getContribution<InlineEditController>(InlineEditController.ID);
+	}
+
+	private _currentEdit: ISettableObservable<InlineEditWidget | undefined> = disposableObservableValue(this, undefined);
```
[the remark this answers is not in the document]
Comment (Krzysztof-Cieslak) [reply src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts]: ✅

File: src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts (hunk 8 above)
Comment (Krzysztof-Cieslak) [reply src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts]: Oh, I've realized we don't need this at all - it's leftover from one of our experiments where we were trying to do background colouring also for ghost text. This turned out to be too heavy/distracting. 

I removed this change to minimize diff.

File: src/vs/workbench/api/common/extHostTypes.ts (hunk 12 above)
Comment (Krzysztof-Cieslak) [reply src/vs/workbench/api/common/extHostTypes.ts:4311]: Updated the interface to make `text` and `range` read-only.

**Comment by Krzysztof-Cieslak** [comment]
@hediet I've addressed all the feedback

File: src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+			if (!currentEdit) {
+				this._isVisibleContext.set(false);
+				this._isCursorAtInlineEditContext.set(false);
+				return;
+			}
+			this._isVisibleContext.set(true);
+			this._isCursorAtInlineEditContext.set(false);
+			const pos = editor.getPosition();
+			if (pos) {
+				this.checkCursorPosition(pos);
```
Comment (hediet) [inline src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts:100]: Why is this needed? `onDidChangeCursorPosition` should take care of this

File: src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+		this._register(editor.onDidChangeModelContent(async () => {
+			if (!this._enabled.get()) {
+				return;
+			}
+			await this.getInlineEdit(editor, true);
+		}));
+
+		//Check if the cursor is at the ghost text
+		this._register(editor.onDidChangeCursorPosition((e) => {
+			if (!this._enabled.get()) {
```
Comment (hediet) [inline src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts:81]: Can you only subscribe to `onDidChangeCursorPosition` when enabled is true?
This feature will probably only be enabled for 0.001% of VS Code users and this code is called billions of times per day in the world.

You can do it like this:
```
const cursorPosition = observableFromEvent(editor.onDidChangeCursorPosition, () => editor.getPosition());
this._register(autorun(reader => {
	if (!this.enabled.read(reader)) { return; }

	const pos = cursorPosition.read(reader);
	if (pos) {
		this.checkCursorPosition(pos);
	}
}));
```

File: src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+			this.checkCursorPosition(e.position);
+		}));
+
+		//Perform stuff when the current edit has changed
+		this._register(autorun((reader) => {
+			/** @description InlineEditController.update model */
+			const currentEdit = this._currentEdit.read(reader);
+			if (!currentEdit) {
+				this._isVisibleContext.set(false);
+				this._isCursorAtInlineEditContext.set(false);
```
Comment (hediet) [inline src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts:93]: This is always set to false, at least in both the if and else branch?

File: src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+		@ILanguageFeaturesService private readonly languageFeaturesService: ILanguageFeaturesService,
+		@ICommandService private readonly _commandService: ICommandService,
+		@IConfigurationService private readonly _configurationService: IConfigurationService,
+	) {
+		super();
+
+		//Automatically request inline edit when the content was changed
+		//Cancel the previous request if there is one
+		//Remove the previous ghost text
+		this._register(editor.onDidChangeModelContent(async () => {
```
Comment (hediet) [inline src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts:72]: Can you only subscribe to onDidChangeCursorPosition when enabled is true?

File: src/vs/workbench/api/common/extHostLanguageFeatures.ts [hunk 19]
```diff
@@ -1355,6 +1355,64 @@ class InlineCompletionAdapter extends InlineCompletionAdapterBase {
...
 }
 
+class InlineEditAdapter {
+	private languageTriggerKindToVSCodeTriggerKind: Record<languages.InlineEditTriggerKind, InlineEditTriggerKind> = {
+		[languages.InlineEditTriggerKind.Automatic]: InlineEditTriggerKind.Automatic,
+		[languages.InlineEditTriggerKind.Invoke]: InlineEditTriggerKind.Invoke,
+	};
+
+	async provideInlineEdits(uri: URI, context: languages.IInlineEditContext, token: CancellationToken): Promise<languages.IInlineEdit | undefined> {
+		let disposableStore: DisposableStore | undefined = undefined;
```
Comment (hediet) [inline src/vs/workbench/api/common/extHostLanguageFeatures.ts:1365]: `disposableStore` is leaking! Look at how inline completions deal with this:

```
const pid = this._references.createReferenceId({
	dispose() {
		disposableStore?.dispose();
	},
	items: normalizedResult
});
```
Unfortunately, getting this right adds quite a lot of complexity...

**Review by hediet — COMMENTED** [review]
See comments. Looks okay, just the leaking disposable store is a blocker at the moment.

File: src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts (hunk 15 above)
Comment (Krzysztof-Cieslak) [reply src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts:100]: It's there to handle a case where the user has moved the cursor while we're getting the next suggestion, and in a moment when we display the next suggestion, the user's cursor is accidentally on the line of suggestion.

File: src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts (hunk 17 above)
Comment (Krzysztof-Cieslak) [reply src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts:93]: It can be set to true in `checkCursorPosition` - but yeah, I've done some minor cleanup here.

File: src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts (hunk 16 above)
Comment (Krzysztof-Cieslak) [reply src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts]: I've made the change you suggested here, but I'm not 100% sure how it makes things better - I've thought `autorun` would be rerun if any observed `IObservable` has changed so changing `cursorPosition` (which happens on every cursor position change) would trigger this `autorun`... so it's exactly same behaviour as before?

File: src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts (hunk 18 above)
Comment (Krzysztof-Cieslak) [reply src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts]: ✅ (Tho, check comment on `onDidChangeCursorPosition`)

File: src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts (hunk 16 above)
Comment (Krzysztof-Cieslak) [reply src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts]: Oh, that makes sense! Thanks for explanation.

**Review by chrmarti — APPROVED** [review]
